### ISSUE TITLE
scylla 3.1.1, records disappear in big table after upgrad from 3.0

### ISSUE THREAD

**Issue by Termite** [scylladb/scylla#6040]
Hi,

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 3.1.1
Cluster size: 8
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian stretch

Platform (physical/VM/cloud instance type/docker): physical
Hardware: sockets= 1, cores=4, hyperthreading=8, memory=64GB
Disks: HDD, 2 per node as raid-0,  ca 7TB per node

The discussion to this issue is here:
https://groups.google.com/d/msgid/scylladb-users/c953b65c-b718-c789-5626-e5ee9beb1560%40fantasymail.de

We are facing data loss after upgrading from 3.0 to 3.1.1
This table has around 6.000.000.000 records, each consting of a few short blobs (< 60 bytes),
some numbers, a text field, a timestamp and a map<text, frozen<set<text>>>, which is null at
the moment.

After the import we made a checked to be sure all data is imported.
We have some tests running regulary and didn't notice any data missing.

After the upgrade to version 3.1.1.  records started to disappear. 
We looked for bugs in our software, changed things, reimported, more than once. 
It didn't help, there are allways rows missing which were imported with version 3.0.

After we got a second cluster for testing, we installed 3.0 on it, imported the data and let it run for four weeks to see what happens. The data on 3.0 was complete and stayed complete.
Last Friday I made the upgrade to 3.1.1 and just after the update records disappeared.
Out of a testset of 25000 records, around 5000 are missing, that's 25%.

I now think, that there might indeed be a subtle bug in the database.

To help debugging this, I made a data generator in go, which generates data which looks exactly like out data. 

I will help as much as I can, to find the problem.

The next step for us might be to upgrade to 3.2.4 and make another import and check if that works.
This would give a frame to search for a possible bug.

cheers,
 Michael
 
[generator.zip](https://github.com/scylladb/scylla/files/4349713/generator.zip)

**Comment by slivne** [comment]
@Termite 
- Thanks, we will start working on reproducing the issue with what you provided, if there is a bug we want to find it and fix it.
- Just so its clear were you able to reproduce the issue with the reproducer you provided ?

**Comment by Termite** [comment]
Am Freitag, 20. März 2020, 00:43:17 CET schrieb Shlomi Livne:
> @Termite
> - Thanks, we will start working on reproducing the issue with what you
> provided, if there is a bug we want to find it and fix it. 
> - Just so its
> clear were you able to reproduce the issue with the reproducer you provided
> ?

Thanks for the answer.

We have so far tested with our real data, not with data from the generator. 
But the data from the generator is not distinguishable from our real data.

We would have to make a complete dump again and check every record and at the 
moment we can't do that.

I don't know if clustersize play a role in the bug, the genreal test would be 
to call the generator with "go run ./generator.go -records 1200000000 > 
dumpfile.json" and import that. Then make dump and check if all records are 
there. Maybe let random selects run for some time . Then make the update to 
3.1.1 and let checks run again, or dump all data again and compare with the 
inital import file. Quite time consuming dependend on hardware.

We use smaller import files to make it easier to handle things.

 Michael

**Comment by slivne** [comment]
@eliransin please update on progress here - so it will be clear what you have done so far and where we stand

**Comment by eliransin** [comment]
Hi @Termite , I am documenting here the details I have so far and the information I have so you can correct me if I am wrong.
I use the following schema:
```
CREATE KEYSPACE test
  WITH REPLICATION = { 
   'class' : 'NetworkTopologyStrategy', 
   'us-east' : 3
  } ;

CREATE TABLE test.errorness_table (
key blob,
id blob,
attr map<text, frozen<set<text>>>,
blob1 blob,
blob2 blob,
blob3 blob,
blob4 blob,
ts timestamp,
text1 text,
text2 text,
text3 text,
int1 bigint,
int2 bigint,
primary key (key, id)) WITH compaction = {
'class':'SizeTieredCompactionStrategy'}
AND compression = { 'sstable_compression':'org.apache.cassandra.io.compress.LZ4Compressor' , 'chunk_length_kb': '16'};
```
The generator also needed a little change so it formats the timestamp field so that Scylla 
can understand parse it.
attached is the revised generator:
[generator.zip](https://github.com/scylladb/scylla/files/4374263/generator.zip)

I was doing the following experiment:
1. created a 3.0.11 3 node aws cluster (`i3.8xlarge` nodes).
2. used: `go run generator.go --records 100000` on every node to insert **400000** records.
3. validated that all of the data was there using the following cqlsh command: `select count(*) from test.errorness_table` it yielded `1200000`
4. performed a rolling upgrade procedure as it is defined [here](https://docs.scylladb.com/upgrade/upgrade-opensource/upgrade-guide-from-3.0-to-3.1/upgrade-guide-from-3.0-to-3.1-rpm/) with the slight change that I used the following yum command after changing the repo: `yum install scylla-3.1.1 scylla-ami-3.1.1`
5. did step 3 again and got **the same** number of records.

I would like to ask you to review it and give your opinion if the procedure should have restored the problem. If you identify something wrong with the reproduction procedure please let me know.

Next step for me is to do a full scan of the data and count it myself and query the keys one by one to
see if I can find all of them.
If you have a more concrete reproducer I will be glad to try it out.

Thanks,

**Comment by eliransin** [comment]
Also "SELECT *" on the table retrieves all records.

**Comment by Termite** [comment]
Thanks for the tests.
I don't know if cluster size plays a role, we have 8 nodes.

And, since we imported quite a few times, we know that with few data the error
doesn't show up. So you need to import at least 5*10^9 rows to see it.

Also, I don't know the period of the random number generator of the importer.
So with that amount of data, hopefully there are no duplicates if you run it
in parallel.


 Michael

**Comment by eliransin** [comment]
@Termite one more question, did you have mc format enabled in 3.0 prior to the upgrade?
Can you send me the `scylla.yaml` configuration file you used for the nodes?

**Comment by Termite** [comment]
Am Dienstag, 7. April 2020, 08:22:18 CEST schrieb Eliran Sinvani:
> @Termite one more question, did you have mc format enabled in 3.0 prior to
> the upgrade? Can you send me the `scylla.yaml` configuration file you used
> for the nodes?

Yes, we enabled mc format in 3.0.

cheers
 Michael

**Comment by eliransin** [comment]
@Termite is it possible for you to send me the `scylla.yaml` that you used in the nodes?
I am trying to reproduce and it is better to do it with the exact same configuration you used.

**Comment by Termite** [comment]
Am Sonntag, 12. April 2020, 07:20:07 CEST schrieb Eliran Sinvani:
> @Termite is it possible for you to send me the `scylla.yaml` that you used
> in the nodes? I am trying to reproduce and it is better to do it with the
> exact same configuration you used.

of course. I sent it on 08.04.2020, did the attachment  got lost?

Here it is again, just mail me if it's ok.

Cheers
 Michael

**Comment by eliransin** [comment]
@Termite I believe the problem is that github doesn't "understand" attachments through emails.
Can you please try to attach it through the github gui? (in the browser).
Thanks a lot,

**Comment by Termite** [comment]
Ok, didn't  know that. Here is the config..
[scylla.yaml.txt](https://github.com/scylladb/scylla/files/4466231/scylla.yaml.txt)

**Comment by eliransin** [comment]
@Termite thanks, I went through ok now :slightly_smiling_face:

**Comment by Termite** [comment]
Hi,
a little update:  we imported the data in a 3.2 version over the weekend.
First we got timeout exceptions, which we did not get with version 3.1, so we reduced the number of import threads, then the exceptions disappeared. 
But shortly after the import there was missing data.

So, we will upgrade to version 3.3 and import again, this time with cdc enabled, to see where data is changed. This should take 4 days at least to complete.

cheers
 Michael

**Comment by eliransin** [comment]
> Hi,
> a little update: we imported the data in a 3.2 version over the weekend.

did you use the method you described before? A java process that read a csv/json and does a number of parallel inserts with QUORUM cl?

> First we got timeout exceptions, which we did not get with version 3.1, so we reduced the number of import threads, then the exceptions disappeared.

Did you restart the import on a clean cluster after changing the thread count or did you simply continue from where you left off?

> But shortly after the import there was missing data.
> 

Was the data "seen" before it disappeared (a select query returned it) or could it be that it failed on the insert stage?

> So, we will upgrade to version 3.3 and import again, this time with cdc enabled, to see where data is changed. This should take 4 days at least to complete.
> 
> cheers
> Michael

Thank you very much for the update, we are now trying to reproduce.
Your answers for the above questions will be much appreciated.

Just to see that I understand the bottom line, you didn't do any upgrade, you just installed fresh 3.2 on a new cluster and ran an import process which is a massive insert with a java program into the database and after it was completed there was data missing. correct?

**Comment by eliransin** [comment]
@Termite , one more request if I may, is it possible to get your java import client. We try to match your 
process as much as possible so we increase the chance of reproducing this issue.

**Comment by Termite** [comment]
Hi,

here are the answers..

Am Sonntag, 19. April 2020, 09:21:01 CEST schrieb Eliran Sinvani:
>
> did you use the method you described before? A java process that read a
> csv/json and does a number of parallel inserts with QUORUM cl?

Yes we did.

> Did you restart the import on a clean cluster after changing the thread
> count or did you simply continue from where you left off?

We've imported to a clean cluster, right from the beginning. We did use
a freshly created table.

>
> Was the data "seen" before it disappeared (a select query returned it) or
> could it be that it failed on the insert stage?

We have seen this during the latest imports but we did see this
behaviour during earlier import runs. We' did also see that data went
missing after the import was finished during repair runs. (Data once
returned by selects on the csql shell was gone some minutes later)


> Just to see that I understand the bottom line, you didn't do any upgrade,
you
> just installed fresh 3.2 on a new cluster and ran an import process which is
a
> massive insert with a java program into the database and after it was
> completed there was data missing. correct?

We installed in a 3.0: Result ok
Upgrade to 3.1.1.: Result nok
Upgrade to 3.2: Result nok
Upgrade to 3.3: Result nok

It seems, stuff was broken between 3.0 and 3.1.1


Thanks for your help, cheers
 Michael,

**Comment by Termite** [comment]
So after importing into version 3.3 the data loss is reproducible.
We tried with cdc enabled but the insert rate dropped from ca 15000 per second to less than 200 per second, so it was unusable.

The question on the java code is undecided since there more dependencies

**Comment by eliransin** [comment]
Thanks for all of the information @Termite , I am just not clear on one point, the recent data loss scenario on 3.2:
Did you just installed fresh 3.2 cluster and then imported the data into this cluster?

The reason for the questions is that I am trying to figure out the simplest reproducer possible.

Thanks,

**Comment by eliransin** [comment]
@haaawk FYI
https://github.com/scylladb/scylla/issues/6040#issuecomment-618915968

**Comment by Termite** [comment]
Hi,
as we use debian, we upgrade via dpkg to new scylla versions.
The import was made into a new table, every time we imported.

As data started to disappear after importing, I think there may be an issue with compaction / repair, maybe in conjunction with the number of rows, as the data loss happens only after importing more than around 5x10^9 rows,  The number 5x10^9 is  above 4x10^9, an unsigned 32bit number. 
And, with 3.0 we didn't had data loss. Just my observations.

Thanks,
 Michael

**Comment by eliransin** [comment]
> as the data loss happens only after importing more than around 5x10^9 rows, The number 5x10^9 is above 4x10^9, an unsigned 32bit number.

@Termite thanks for all of the information, your insights are priceless.
We will look into this as well as try to reproduce.
We will keep you updated with our findings.

**Comment by Termite** [comment]
Hi,

I made a go importer which imports data with a calculated key to make checks easier.
The go script can also check rows for existence, and has another mode to constantly check while inserting.

I imported 1000000 rows.
Then I checked those one millions rows, all were found.
Then I started the tool to import 6000000000 rows while constantly checking the existence of the first million of rows.
During the import the first million rows were checked around 6000 times and no missing row was reported.

But:
checking rows *beyond* the first million (after the import) at once showed rows missing.
So I checked the first 1010000 rows (10000 after the first million), over 1000 rows were missing.
Checking the whole range of 6000000000 rows showed a great amount of rows missing, across the whole range of row ids.

This is an interessting result:  does the constantly selecting of the first million rows make them stay in some cache and prevent them from being  "lost" ?

I attached the go script, so that you can reproduce this test:

first import 1000000 rows:
./debug-importer -hosts=scylla1...  -count=100000 -threads=100 -mode=insert

second, check those rows:
./debug-importer -hosts=scylla1...  -count=100000 -threads=100 -mode=check

then import the rest and constantly check the first million:
./debug-importer -hosts=scylla1..   -start=1000000  -count=6000000000 -threads=100 -mode=insert-check

this should take some time. After the import you can check the the rows:
./debug-importer -hosts=scylla1...  -count=1010000 -threads=100 -mode=check
or
./debug-importer -hosts=scylla1...  -count=6000000000 -threads=100 -mode=check

and see what happens.

Michael

[scylla-debugimporter.go.txt](https://github.com/scylladb/scylla/files/4592586/scylla-debugimporter.go.txt)

**Comment by Termite** [comment]
Sorry, there is a typo in the first 2 command lines, the -count should be -count=1000000:
./debug-importer -hosts=scylla1... -count=1000000 -threads=100 -mode=insert
./debug-importer -hosts=scylla1... -count=1000000 -threads=100 -mode=check

**Comment by denesb** [comment]
On Thu, 2020-05-07 at 04:20 -0700, Michael Wohlwend wrote:
> Hi,
> 
> I made a go importer which imports data with a calculated key to make
> checks easier.
> 
> The go script can also check rows for existence, and has another mode
> to constantly check while inserting.
> 
> I imported 1000000 rows.
> 
> Then I checked those one millions rows, all were found.
> 
> Then I started the tool to import 6000000000 rows while constantly
> checking the existence of the first million of rows.
> 
> During the import the first million rows were checked around 6000
> times and no missing row was reported.
> 
> But:
> 
> checking rows beyond the first million (after the import) at once
> showed rows missing.
> 
> So I checked the first 1010000 rows (10000 after the first million),
> over 1000 rows were missing.
> 
> Checking the whole range of 6000000000 rows showed a great amount of
> rows missing, across the whole range of row ids.
> 
> This is an interessting result:  does the constantly selecting of the
> first million rows make them stay in some cache and prevent them from
> being  "lost" ?

Michael, I don't remember whether we already asked this, but does using
'BYPASS CACHE' in the query change the result, i.e. are the lost rows
present when this is used?

**Comment by Termite** [comment]
I think this was not asked. I can check this, it requires a fresh import, which I can start today.

Cheers
 Michael

**Comment by eliransin** [comment]
@Termite just a quick update, @avikivity suggested to first follow your hunch and it seams to be true,
I inserted 6*10^9 rows of a simple table (only a bigint as a partition key) and records are missing
(counting partition keys with sstabledump gave me 17944355893 and it should have been 18000000000 (because of the RF=3 factor) which means lose of 18548036 records (not rows)
I will now try to repair the table and run the count again.
/cc @avikivity

**Comment by slivne** [comment]
@Eliran Sinvani <eliransin@scylladb.com> did you do a nodetool flush ?

**Comment by avikivity** [comment]
Good catch @slivne.

**Comment by eliransin** [comment]
@slivne no, I forgot :flushed: . I will do it after I finish the repair round.

**Comment by eliransin** [comment]
another update - after nodetool flush,repair and compact cycles on all of the nodes, I get a total of **18001099735** records which is slightly larger than the number of keys probably because of compaction method , no I will list the 
actual keys and see if all appear (this is just the sequence of integer between 0 and 6*10^9-1.
/cc @slivne @avikivity

**Comment by avikivity** [comment]
Probably duplicate records. I recommend to use COPY FROM or something 
similar instead.

**Comment by eliransin** [comment]
I realized that I didn't specified how did I insert the data so here it is:
I used this super simple insert program (a python script): 
[data_inserter.txt](https://github.com/scylladb/scylla/files/4607631/data_inserter.txt)
I ran it on 4 loaders with 200 threads with 4 ranges (0-1500000000,1500000000-3000000000,
3000000000-4500000000,4500000000-6000000000)
The last line for all runs reported a success for all records.

I have tried to count the rows using the copy to command and this is what I found:
```
cqlsh> copy test.pk_only to '/var/lib/scylla/keys/keys.csv';
Using 7 child processes

Starting copy of test.pk_only with columns [pk].
Processed: 5822810301 rows; Rate:   53785 rows/s; Avg. rate:  396359 rows/s
5822810301 rows exported to 1 files in 4 hours, 4 minutes, and 50.810 seconds.
cqlsh> quit
[centos@eliransin_6040_debug-eliransin-db-node-f078b4ca-4 ~]$ cat /var/lib/scylla/keys/keys.csv | wc -l
5822810301
[centos@eliransin_6040_debug-eliransin-db-node-f078b4ca-4 ~]$
```

Next steps are:
1. Find a record that doesn't appear in the csv and search for it in the actual database.
2. Do the same experiment but use the COPY program to insert the data as well as to extract and
count the data. Do the extraction before and after each step (flush,repair,compact).

**Comment by eliransin** [comment]
So far I found 2 distinct problems (from my first insert):
1. An entirely missing record.
2.  I found 1 record which probably wasn't present on all replicas,because it wasn't found with the default consistency (ONE) but was present with CL=ALL. It doesn't make sense since I ran repair on all of the nodes.

I am going to do the second insert now with COPY UTIL. 
This one is going to be with CL=ONE.
@Termite Just making sure, do you have error recovery in your java code for case a record failed to be inserted?

**Comment by eliransin** [comment]
@Termite another question (which might have been asked already). What consistency level did you use in the imports and what consistency level did you use for the check (of the first 1000000 rows)?

**Comment by Termite** [comment]
Thanks for making this test.
Yes, we have complete error logging and retrying logic in the java code.
The go script I attached also logs errors.

Consistency is allways quorum.

Did you check the go script I uploaded last week and the test I ran? 
See https://github.com/scylladb/scylla/issues/6040#issuecomment-625193557

There you can see the quorum setting in the connect function.
I descriped my test with the go script, so you can make the exact same test, this would make the results (ours and yours) a little more comparable.

I have started a new import last friday with using "bypass cache" in the select statement to check the first million of records. This import finished yesterday and I'll check today if using "bypass cache" makes a difference.

**Comment by Termite** [comment]
I just checked the import where I used "bypass cache" to constantly verify the first million rows.
This has indeed an effect, there are now a lot of rows missing, which was not the case without bypass cache.
I suggest to use my go script to make the check again. Also, the use case of a DB  is "insert", then "select" to get the data.  sstabledump or other methods might deliver other results, but then you cannot directly compare it.

 Michael

**Comment by eliransin** [comment]
> Did you check the go script I uploaded last week and the test I ran?
> See [#6040 (comment)](https://github.com/scylladb/scylla/issues/6040#issuecomment-625193557)

Sorry @Termite I must have missed it. I will make sure to run your test when I finish mine (if we don't find the problem now)
Thank you for taking the time to help us with debugging this issue :pray:  , we don't take your efforts for granted and we are very serious about getting this solved.

**Comment by eliransin** [comment]
A short update:
1. I have inserted the data (6*10^9 records) and read it right afterwards:
```
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_cl_one.csv' with MAXATTEMPTS = 1000000; 
Export - CL ONE
----------------
Starting copy of test.pk_only4 with columns [pk].
Processed: 6000000000 rows; Rate:   55845 rows/s; Avg. rate:  387711 rows/svg. rate:  393311 rows/s
6000000000 rows exported to 1 files in 4 hours, 17 minutes, and 55.543 seconds.
```
2. Did a rolling restart of the nodes and major compaction right after and read it again:
```
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_cl_after_compaction.csv' with MAXATTEMPTS = 1000000; 
Export - After compaction
----------------------------
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_cl_after_compaction.csv' with MAXATTEMPTS = 1000000; 
Using 7 child processes

Starting copy of test.pk_only4 with columns [pk].
Processed: 5824454853 rows; Rate:   61998 rows/s; Avg. rate:  376499 rows/s
5824454853 rows exported to 1 files in 4 hours, 17 minutes, and 50.098 seconds.
```
and also ran this just in case COPY messes up with the stats:
```
[centos@eliransin_6040_debug-eliransin-db-node-f078b4ca-4 keys]$ cat cp_read_cl_after_compaction.csv | wc -l
5824454853
```
So according to this we lost ~170M records.

I will now further brake it down, next steps (validating the record count after each step):
```
1. reinsert data
2. nodetool flush
3. Clear cache (@slivne/ @avikivity   can you remind me how to do this?)
4. major compaction
5.repair
6. major compaction
5.rolling restart
``
I am hoping that we will be able to pin the mechanism that "losses" the data with this steps:
CommitLogs/Compaction/Repair

**Comment by eliransin** [comment]
BTW I will also try next to find a smaller reproducer so we can further debug this:
If repair is not the problem maybe a one node cluster with 2.5*10^9 records is enough since this is the
amount of records on each one of the nodes. If this also reproduces using this we can reduce the amount of records gradually until we have the smallest reproducer.

**Comment by avikivity** [comment]
With CL=ONE you can't expect consistent data, in case there were problems during the write (there shouldn't be, but it's not guaranteed), You need to read with LOCAL_QUORUM to be sure.

**Comment by avikivity** [comment]
read AND write

**Comment by eliransin** [comment]
I know that that was a mistake on my part this is why I documented it, I will make sure I do it on this run. But in this specific case the first read (just after the write) retrieved all of the records so I guess they persisted somewhere right?

**Comment by eliransin** [comment]
@avikivity Doing the new test with QUORUM.

**Comment by avikivity** [comment]
> I know that that was a mistake on my part this is why I documented it, I will make sure I do it on this run. But in this specific case the first read (just after the write) retrieved all of the records so I guess they persisted somewhere right?

You can have the first read access a replica that had data, and a second read access a replica that did not have the data.

**Comment by avikivity** [comment]
Do you have metrics? this should show as background reads failed.

**Comment by slivne** [comment]
clearing the cache

```
  curl -X POST http://localhost:10000/lsa/compact
```

**Comment by eliransin** [comment]
> Do you have metrics? this should show as background reads failed.

I do have metrics but I prefer to do it right than to dig in them:slightly_smiling_face: 

Is the plan for pinning down the problem sounds reasonable?

**Comment by eliransin** [comment]
An update about the reproduction, it looks like that compaction loses rows,
I am now working to identify an actual missing key/keys so we can try and understand if
it is somehow connected to the actual keys or just sporadic loses in compation, all queries
done with CL LOCAL_QUORUM:
```
CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication': '3'}  AND durable_writes = true;
CREATE TABLE test.pk_only4 (pk bigint PRIMARY KEY); 

Last attempt
---------------------
cqlsh> CONSISTENCY LOCAL_QUORUM
Consistency level set to LOCAL_QUORUM.
cqlsh>  COPY test.pk_only4 (pk) FROM '/var/lib/scylla/keys/cp_insert.csv' with MAXATTEMPTS = 1000000000 AND MAXINSERTERRORS=-1;
Using 7 child processes
Starting copy of test.pk_only4 with columns [pk].
No records inserted in 90 seconds, abortingows/s; Avg. rate:   70823 rows/s
Processed: 5999975000 rows; Rate:   31446 rows/s; Avg. rate:   70748 rows/s
5999975000 rows imported from 1 files in 23 hours, 33 minutes, and 28.126 seconds (0 skipped).

Right after the write :)
-------------------------
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_after_write.csv' with MAXATTEMPTS = 1000000; 
Using 7 child processes

Starting copy of test.pk_only4 with columns [pk].
Processed: 6000000000 rows; Rate:   47247 rows/s; Avg. rate:  331935 rows/s
6000000000 rows exported to 1 files in 5 hours, 1 minute, and 15.920 seconds.

COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_after_flush.csv' with MAXATTEMPTS = 1000000; 

After flush
----------------
COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_after_flush.csv' with MAXATTEMPTS = 1000000; 
Using 7 child processes

Starting copy of test.pk_only4 with columns [pk].
Processed: 6000000000 rows; Rate:   48993 rows/s; Avg. rate:  389025 rows/s
6000000000 rows exported to 1 files in 4 hours, 17 minutes, and 3.244 seconds.


After clear cash
----------------------
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_after_clear_cache.csv' with MAXATTEMPTS = 1000000; 
Using 7 child processes

Starting copy of test.pk_only4 with columns [pk].
Processed: 6000000000 rows; Rate:   44366 rows/s; Avg. rate:  394619 rows/s
6000000000 rows exported to 1 files in 4 hours, 13 minutes, and 24.623 seconds.

After Compaction
-----------------
cqlsh> COPY test.pk_only4 (pk) TO '/var/lib/scylla/keys/cp_read_after_compaction.csv' with MAXATTEMPTS = 1000000; 
Using 7 child processes

Starting copy of test.pk_only4 with columns [pk].
Processed: 5916012423 rows; Rate:   52808 rows/s; Avg. rate:  336155 rows/s
5916012423 rows exported to 1 files in 4 hours, 53 minutes, and 19.160 seconds.
cqlsh> 

[centos@eliransin_6040_debug-eliransin-db-node-f078b4ca-4 keys]$ wc -l cp_read_after_compaction.csv 
5916012423 cp_read_after_compaction.csv
```
/cc @avikivity @slivne

**Comment by eliransin** [comment]
OK - I have a very strange result:
First Of All a translation table:

node # | priv ip | global ip| long name
|---|---|---|---
|1| 172.30.0.158| 54.84.234.41|eliransin_6040_debug-eliransin-db-node-f078b4ca-1
|2| 172.30.0.169 | 54.164.7.236|eliransin_6040_debug-eliransin-db-node-f078b4ca-2
|3|172.30.0.218 |3.87.226.44|eliransin_6040_debug-eliransin-db-node-f078b4ca-3
|4|172.30.0.212 |54.165.251.218|eliransin_6040_debug-eliransin-db-node-f078b4ca-4
|5|172.30.0.128 |54.235.226.51|eliransin_6040_debug-eliransin-db-node-f078b4ca-5
|6|172.30.0.159 |54.159.119.209|eliransin_6040_debug-eliransin-db-node-f078b4ca-6
|7|172.30.0.190 |3.86.43.142|eliransin_6040_debug-eliransin-db-node-f078b4ca-7
|8|172.30.0.163 |3.83.46.112|eliransin_6040_debug-eliransin-db-node-f078b4ca-8

We have two keys:
0 - which can be found via query
77 - which can't be found via query (but only after compaction)
```
cqlsh> CONSISTENCY ALL
Consistency level set to ALL.
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 0;

 pk
----
  0

(1 rows)
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 77;                                                                                                                                                                                             

 pk
----

(0 rows)

```
First I tried to check where the key is **expected** to be found:
```
nodetool getendpoints test pk_only4 0
172.30.0.169 (node 2)
172.30.0.163 (node 8)
172.30.0.218 (node 3)

nodetool getendpoints test pk_only4 77
172.30.0.169 (node 2)
172.30.0.128 (node 5)
172.30.0.218 (node 4)

using:
curl "localhost:10000/storage_service/natural_endpoints/test?cf=pk_only&key=0"
curl "localhost:10000/storage_service/natural_endpoints/test?cf=pk_only&key=77"

returns the exact same results (unsurprisingly since nodetool is using this API :slightly_smiling_face: )
```

Now in **each** node I ran the following command to search for the key in the sstables - even the ones that are not expected to have the key:
```
"cd /var/lib/scylla/data && find test -name "*Data.db" | grep pk_only4 | xargs -n 1 -i sstabledump {} -k <key> 2>/dev/null"
explenation: 
1. I do it in a loop through ssh so I always start at ~ dir, this is what the cd ... is for.
2. The 2>/dev/null part gets read of the anoying cdc header message so I can get a clean log.
```

I got from doing this:
```
for ip in `cat 6040_db_ips`;do echo $ip; ssh -i ~/.ssh/scylla-qa-ec2 centos@$ip "hostname && cd /var/lib/scylla/data && find test -name "*Data.db" | grep pk_only4 | xargs -n 1 -i sstabledump {} -k 0 2>/dev/null"; echo ""; done
54.84.234.41
eliransin_6040_debug-eliransin-db-node-f078b4ca-1
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
54.164.7.236
eliransin_6040_debug-eliransin-db-node-f078b4ca-2
[ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "0" ],
      "position" : 5935232622
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 5935232650,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.722340Z" },
        "cells" : [ ]
      }
    ]
  }
][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
3.87.226.44
eliransin_6040_debug-eliransin-db-node-f078b4ca-3
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
54.165.251.218
eliransin_6040_debug-eliransin-db-node-f078b4ca-4
[ ][ ][ ][ ][ ][ ][ ][ ]
54.235.226.51
eliransin_6040_debug-eliransin-db-node-f078b4ca-5
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "0" ],
      "position" : 5745940760
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 5745940788,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.722340Z" },
        "cells" : [ ]
      }
    ]
  }
][ ][ ][ ][ ]
54.159.119.209
eliransin_6040_debug-eliransin-db-node-f078b4ca-6
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
3.86.43.142
eliransin_6040_debug-eliransin-db-node-f078b4ca-7
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
3.83.46.112
eliransin_6040_debug-eliransin-db-node-f078b4ca-8
[ ][ ][ ][ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "0" ],
      "position" : 5541604078
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 5541604106,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.722340Z" },
        "cells" : [ ]
      }
    ]
  }
]
[eliransin@localhost ~]$ for ip in `cat 6040_db_ips`;do echo $ip; ssh -i ~/.ssh/scylla-qa-ec2 centos@$ip "hostname && cd /var/lib/scylla/data && find test -name "*Data.db" | grep pk_only4 | xargs -n 1 -i sstabledump {} -k 77 2>/dev/null"; echo ""; done
54.84.234.41
eliransin_6040_debug-eliransin-db-node-f078b4ca-1
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
54.164.7.236
eliransin_6040_debug-eliransin-db-node-f078b4ca-2
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
3.87.226.44
eliransin_6040_debug-eliransin-db-node-f078b4ca-3
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "77" ],
      "position" : 8517898641
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 8517898669,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.632133Z" },
        "cells" : [ ]
      }
    ]
  }
][ ][ ][ ]
54.165.251.218
eliransin_6040_debug-eliransin-db-node-f078b4ca-4
[ ][ ][ ][ ][ ][ ][ ][ ]
54.235.226.51
eliransin_6040_debug-eliransin-db-node-f078b4ca-5
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
54.159.119.209
eliransin_6040_debug-eliransin-db-node-f078b4ca-6
[ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "77" ],
      "position" : 8663146648
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 8663146676,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.632133Z" },
        "cells" : [ ]
      }
    ]
  }
][ ][ ]
3.86.43.142
eliransin_6040_debug-eliransin-db-node-f078b4ca-7
[ ][ ][ ][ ][ ][ ][ ][ ][ ][
  {
    "partition" : {
      "key" : [ "77" ],
      "position" : 8679173565
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 8679173593,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.632133Z" },
        "cells" : [ ]
      }
    ]
  }
][ ][ ][ ][ ][ ][ ]
3.83.46.112
eliransin_6040_debug-eliransin-db-node-f078b4ca-8
[ ][ ][ ][ ][ ][ ][ ][ ]
```

What we can see here is that 0 is on nodes 2,**3**,8 while the expected ones are 2,5,8
and 77 is on nodes **3**,**6**,**7** while the expected are 2,4,5

This can explain all of the behaviors that were described by @Termite :
Querying 0 with cosistency level LOCAL_ONE would have given the "appearance" of a record that comes and goes.
Querying 77 with any CL just doesn't return results.

It seams like the real problem is tokens? The endpoints that are read from are not the ones that are
written to - this is assuming that the coordinator node uses the same logic to determine the key endpoints as nodetool/rest api.

Another **IMPORTANT** note : The missing records were retrieved by the COPY utility just before
the compaction.

Next steps:
1. Look into some more keys (non missing and missing).
2. Try to narrow down the possibilities (I am not sure how compaction is affecting this except that maybe before compaction those keys were present on more than 3 nodes and compaction go read of them for some reason).
3. Move to 4.0 and see if the bug is also there.
4. find a smaller use case. (maybe 4 nodes with rf 3 ) - and this time save a snapshot of the sstables right after "nodetool flush".

@Termite it would be of great help if you do the procedure I made and try to see if the missing records are present on any of the nodes (by using some variation sstabledump command i wrote above)

**Comment by eliransin** [comment]
OK - now the queries with tracing shows that get natural endpoints is wrong (which is a different bug???)
Queries with tracing:
```
cqlsh> CONSISTENCY ALL                                                                                                                                                                                                               [73/1970]
Consistency level set to ALL.
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 0;

 pk
----
  0

(1 rows)

Tracing session: c52f20b0-98c9-11ea-ae2d-000000000000

 activity                                                                                                                                                                                      | timestamp                  | source       | s
ource_elapsed | client
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+--------------+--
--------------+--------------
                                                                                                                                                                            Execute CQL3 query | 2020-05-18 05:38:15.227000 | 172.30.0.169 |  
            0 | 172.30.0.212
                                                                                                                                                                 Parsing a statement [shard 2] | 2020-05-18 05:38:15.227798 | 172.30.0.169 |  
           -- | 172.30.0.212
                                                                                                                                                              Processing a statement [shard 2] | 2020-05-18 05:38:15.227848 | 172.30.0.169 |  
           51 | 172.30.0.212
 Creating read executor for token 2945182322382062539 with all: {172.30.0.169, 172.30.0.128, 172.30.0.163} targets: {172.30.0.169, 172.30.0.128, 172.30.0.163} repair decision: NONE [shard 2] | 2020-05-18 05:38:15.227885 | 172.30.0.169 |  
           88 | 172.30.0.212
                                                                                                                                                         read_data: querying locally [shard 2] | 2020-05-18 05:38:15.227888 | 172.30.0.169 |  
           90 | 172.30.0.212
                                                                                                                                     read_digest: sending a message to /172.30.0.128 [shard 2] | 2020-05-18 05:38:15.227893 | 172.30.0.169 |  
           96 | 172.30.0.212
                                                                                                                                     read_digest: sending a message to /172.30.0.163 [shard 2] | 2020-05-18 05:38:15.227904 | 172.30.0.169 |  
          106 | 172.30.0.212
                                                                                                                 Start querying the token range that starts with 2945182322382062539 [shard 7] | 2020-05-18 05:38:15.227954 | 172.30.0.169 |  
           11 | 172.30.0.212
                                                                                                                                                                    Querying is done [shard 7] | 2020-05-18 05:38:15.227977 | 172.30.0.169 |  
           33 | 172.30.0.212
                                                                                                                                        read_digest: got response from /172.30.0.163 [shard 2] | 2020-05-18 05:38:15.228386 | 172.30.0.169 |  
          588 | 172.30.0.212
                                                                                                                                    read_digest: message received from /172.30.0.169 [shard 2] | 2020-05-18 05:38:15.229407 | 172.30.0.163 |  
           11 | 172.30.0.212
                                                                                                                 Start querying the token range that starts with 2945182322382062539 [shard 7] | 2020-05-18 05:38:15.229444 | 172.30.0.163 |  
            9 | 172.30.0.212
                                                                                                                                                                    Querying is done [shard 7] | 2020-05-18 05:38:15.229460 | 172.30.0.163 |  
           26 | 172.30.0.212
                                                                                                                   read_digest handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:38:15.229592 | 172.30.0.163 |  
          197 | 172.30.0.212
                                                                                                                                    read_digest: message received from /172.30.0.169 [shard 2] | 2020-05-18 05:38:15.232651 | 172.30.0.128 |  
         2305 | 172.30.0.212
                                                                                                                 Start querying the token range that starts with 2945182322382062539 [shard 7] | 2020-05-18 05:38:15.232729 | 172.30.0.128 |  
           16 | 172.30.0.212
                                                                                                                                                                    Querying is done [shard 7] | 2020-05-18 05:38:15.232756 | 172.30.0.128 |  
           42 | 172.30.0.212
                                                                                                                   read_digest handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:38:15.235548 | 172.30.0.128 |  
                                                                                                                   read_digest handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:38:15.235548 | 172.30.[23/1970]
         5202 | 172.30.0.212
                                                                                                                                        read_digest: got response from /172.30.0.128 [shard 2] | 2020-05-18 05:38:15.240806 | 172.30.0.169 |  
        13008 | 172.30.0.212
                                                                                                                                                Done processing - preparing a result [shard 2] | 2020-05-18 05:38:15.240828 | 172.30.0.169 |  
        13031 | 172.30.0.212
                                                                                                                                                                              Request complete | 2020-05-18 05:38:15.240039 | 172.30.0.169 |  
        13039 | 172.30.0.212


cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 77;                                                                                                                                                                                             

 pk
----

(0 rows)

Tracing session: 3cba75d0-98ca-11ea-ae2d-000000000000

 activity                                                                                                                                                                                          | timestamp                  | source      
 | source_elapsed | client
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+-------------
-+----------------+--------------
                                                                                                                                                                                Execute CQL3 query | 2020-05-18 05:41:35.789000 | 172.30.0.169
 |              0 | 172.30.0.212
                                                                                                                                          read_data: message received from /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.789307 | 172.30.0.190
 |             11 | 172.30.0.212
                                                                                                                                                                     Parsing a statement [shard 2] | 2020-05-18 05:41:35.789342 | 172.30.0.169
 |             -- | 172.30.0.212
                                                                                                                     Start querying the token range that starts with 8790833928959069864 [shard 7] | 2020-05-18 05:41:35.789350 | 172.30.0.190
 |              8 | 172.30.0.212
                                                                                                                                                                        Querying is done [shard 7] | 2020-05-18 05:41:35.789367 | 172.30.0.190
 |             25 | 172.30.0.212
                                                                                                                                                                  Processing a statement [shard 2] | 2020-05-18 05:41:35.789392 | 172.30.0.169
 |             50 | 172.30.0.212
 Creating read executor for token 8790833928959069864 with all: {172.30.0.190, 172.30.0.159, 172.30.0.218} targets: {172.30.0.190, 172.30.0.159, 172.30.0.218} repair decision: DC_LOCAL [shard 2] | 2020-05-18 05:41:35.789440 | 172.30.0.169
 |             98 | 172.30.0.212
                                                                                                                                           read_data: sending a message to /172.30.0.190 [shard 2] | 2020-05-18 05:41:35.789444 | 172.30.0.169
 |            103 | 172.30.0.212
                                                                                                                                         read_digest: sending a message to /172.30.0.159 [shard 2] | 2020-05-18 05:41:35.789457 | 172.30.0.169
 |            115 | 172.30.0.212
                                                                                                                                         read_digest: sending a message to /172.30.0.218 [shard 2] | 2020-05-18 05:41:35.789461 | 172.30.0.169
 |            120 | 172.30.0.212
                                                                                                                         read_data handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.790376 | 172.30.0.190
 |           1080 | 172.30.0.212
                                                                                                                                              read_data: got response from /172.30.0.190 [shard 2] | 2020-05-18 05:41:35.790878 | 172.30.0.169
 |           1536 | 172.30.0.212
                                                                                                                                        read_digest: message received from /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.795726 | 172.30.0.218
 |           4402 | 172.30.0.212
                                                                                                                                        read_digest: message received from /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.795983 | 172.30.0.159
 |           4153 | 172.30.0.212
                                                                                                                     Start querying the token range that starts with 8790833928959069864 [shard 7] | 2020-05-18 05:41:35.796211 | 172.30.0.218
 |            423 | 172.30.0.212
                                                                                                                                                                        Querying is done [shard 7] | 2020-05-18 05:41:35.796225 | 172.30.0.218
 |            438 | 172.30.0.212
                                                                                                                     Start querying the token range that starts with 8790833928959069864 [shard 7] | 2020-05-18 05:41:35.796658 | 172.30.0.159
 |            592 | 172.30.0.212
                                                                                                                                                                        Querying is done [shard 7] | 2020-05-18 05:41:35.796684 | 172.30.0.159
 |            618 | 172.30.0.212
                                                                                                                       read_digest handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.800048 | 172.30.0.159
 |           8218 | 172.30.0.212
                                                                                                                                            read_digest: got response from /172.30.0.159 [shard 2] | 2020-05-18 05:41:35.804514 | 172.30.0.169
 |          15172 | 172.30.0.212
                                                                                                                       read_digest handling is done, sending a response to /172.30.0.169 [shard 2] | 2020-05-18 05:41:35.816627 | 172.30.0.218
 |          25303 | 172.30.0.212
                                                                                                                                            read_digest: got response from /172.30.0.218 [shard 2] | 2020-05-18 05:41:35.823103 | 172.30.0.169
 |          33762 | 172.30.0.212
                                                                                                                                                    Done processing - preparing a result [shard 2] | 2020-05-18 05:41:35.823128 | 172.30.0.169
 |          33787 | 172.30.0.212
                                                                                                                                                                                  Request complete | 2020-05-18 05:41:35.822793 | 172.30.0.169
 |          33793 | 172.30.0.212


cqlsh> 
```

So.... The data is there but is not retrieved for some reason...

**Comment by eliransin** [comment]
Next step look at the sstables metadata to see maybe something is wrong there.

**Comment by eliransin** [comment]
The metadata seams ok (min and max tokens contains the token searched for). 
also running with bypass cache returned nothing:
```
cqlsh> CONSISTENCY ALL
Consistency level set to ALL.
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 77 BYPASS CACHE;

 pk
----

(0 rows)
```

My next (wild) guess is that maybe the excess number of keys messes up the bloom filter (?)

**Comment by Termite** [comment]
As said earlier, we didn't realize the data loss with version 3.0, maybe this can narrow down the search, dependend on the changes in compaction or repair.

**Comment by eliransin** [comment]
As suggested by @slivne  , we have a (relatively) small reproducer:
A docker which has one of the sstables that contains 77 with smp 1 and identical schema with RF=1.
Importing this table and trying to query this infamous 77 still gets nothing:
```
[root@eecaaf98fcd6 data]# cd test/pk_only4-6fbbd2c098ee11ea8f46000000000000/upload/
[root@eecaaf98fcd6 upload]# cp /home/eliransin/6040_dataloss/* .
[root@eecaaf98fcd6 upload]# nodetool refresh test pk_only4
[root@eecaaf98fcd6 upload]# cqlsh
Connected to  at 172.17.0.2:9042.
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
cqlsh> select * from test.pk_only4 where pk = 77;

 pk
----

(0 rows)
cqlsh> quit 
[root@eecaaf98fcd6 upload]# cd ..
[root@eecaaf98fcd6 pk_only4-6fbbd2c098ee11ea8f46000000000000]# ll
total 9235628
-rw-r--r--. 1 root root   17418906 May 18 10:06 mc-1-big-CompressionInfo.db
-rw-r--r--. 1 root root 4575423316 May 18 10:06 mc-1-big-Data.db
-rw-r--r--. 1 root root         10 May 18 10:06 mc-1-big-Digest.crc32
-rw-r--r--. 1 root root  352924176 May 18 10:06 mc-1-big-Filter.db
-rw-r--r--. 1 root root 4508858529 May 18 10:07 mc-1-big-Index.db
-rw-r--r--. 1 root root      88474 May 18 10:07 mc-1-big-Scylla.db
-rw-r--r--. 1 root root       4551 May 18 10:07 mc-1-big-Statistics.db
-rw-r--r--. 1 root root    2541968 May 18 10:07 mc-1-big-Summary.db
-rw-r--r--. 1 root root        102 May 18 10:07 mc-1-big-TOC.txt
drwxr-xr-x. 2 root root          6 May 18 10:00 staging
drwxr-xr-x. 2 root root          6 May 18 10:08 upload
[root@eecaaf98fcd6 pk_only4-6fbbd2c098ee11ea8f46000000000000]# cd ..
[root@eecaaf98fcd6 test]# cd ..
[root@eecaaf98fcd6 data]# sstabledump test/pk_only4-6fbbd2c098ee11ea8f46000000000000/mc-1-big-D
mc-1-big-Data.db       mc-1-big-Digest.crc32  
[root@eecaaf98fcd6 data]# sstabledump test/pk_only4-6fbbd2c098ee11ea8f46000000000000/mc-1-big-Data.db -k 77
[
  {
    "partition" : {
      "key" : [ "77" ],
      "position" : 8679173565
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 8679173593,
        "liveness_info" : { "tstamp" : "2020-05-14T09:59:01.632133Z" },
        "cells" : [ ]
      }
    ]
  }
]
```
So we are down to the sstable indexing/bloom filter that gets messed up during compaction. Given this I will also follow @Termite lead which pinpoints the version range in which the change was made, if I am not mistaken 3.1 introduced ics.
I will try to see if this sstable contains some more "invisible" keys so we have more than one key to work with.
/cc @avikivity

**Comment by eliransin** [comment]
sstable reproducer can be downloaded using the command:
`gsutil cp gs://upload.scylladb.com/7cf8c51a-8ab4-4798-9248-71dcb754bb66/* .`
/cc @bhalevy

**Comment by eliransin** [comment]
This table contains key 77 which doesn't appear in cql results.

**Comment by eliransin** [comment]
Importing the sstable to a docker image of version 3.2.4 is missing the key.
Importing the sstable to a docker image of version 3.0.0 has the key.
Both are started with smp 1 but 3.3.0 creates two sstables while 3.2.4 creates only one.
/cc @bhalevy

**Comment by slivne** [comment]
@eliransin we need 3.0 to read the sstables with compacting them.

Once done please add that info

**Comment by eliransin** [comment]
I did the same procedure for both versions, both dockers started with smp 1:
```
1. cqlsh:
  CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true ;
  CREATE TABLE test.pk_only4 (pk bigint PRIMARY KEY) WITH compaction = {'class': 'NullCompactionStrategy'}; 
2. supervisorctl stop scylla
3. (only for 3.0 - enable mc format)
4. copy the sstable files into the table folder
5.supervisorctl start scyll
6. SELECT * FROM test.pk_only4 WHERE pk = 77;
```
For 3.0:
```
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 77;

 pk
----
 77

(1 rows)
```
For 3.2.4:
```
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
cqlsh> SELECT * FROM test.pk_only4 WHERE pk = 77;

 pk
----

(0 rows)
```

**Comment by avikivity** [comment]
I think it's unlikely that it's the bloom filter, so it's probably the indexing. I recommend bisecting.

**Comment by avikivity** [comment]
What's the schema for that sstable? `pk int primary key`?

**Comment by avikivity** [comment]
Oh, I see it.

**Comment by eliransin** [comment]
I think it's unlikely that it's the bloom filter, so it's probably the indexing. I recommend bisecting.
:+1: will try to bisect.

**Comment by Termite** [comment]
@eliransin: 
 I run the test you made: getting the endpoints of a missing key gave me nodes 1 , 2 and 5.
Then I run on all 8 hosts 'sstabledump  *Data.db  -k <key> 
The key was indeed found three times (RF=3), in one sstable file on three hosts, but not the hosts which "nodetool getendpoints" returned: I found the key on hosts 1, 6 and 7

(A select in the cqlsh does not find the key)

**Comment by Termite** [comment]
An update after checking a few more keys:
a) there are keys for which nodetool getendpoints gives a result (3 nodes in my case), but sstabledump shows three other nodes where the key is stored,
b) there are keys for which nodetool getendpoints gives a result (3 nodes in my case), but sstabledump gives no result at all. (on all nodes).

**Comment by haaawk** [comment]
> An update after checking a few more keys:
> a) there are keys for which nodetool getendpoints gives a result (3 nodes in my case), but sstabledump shows three other nodes where the key is stored,

I can see few options:
1. Cluster was changed (new node or node removed) and those keys used to belong to nodes that sstabledump shows them on and now they belong to the nodes shown by getendpoints. The data hasn't been streamed yet. Not sure how probable this is though.

2. Those sstables with keys that belong to different nodes may be a leftovers after the change of the cluster. If I'm not mistaken, the change of the cluster does the following:
a. stream data from old replicas to new replicas
b. when sstables are compacted on old replicas, data that does not belong to the node any more is removed.
There can be a long time between a and b. Maybe what you're seeing are such old sstables that hasn't been cleaned up after the change of the cluster yet.

> b) there are keys for which nodetool getendpoints gives a result (3 nodes in my case), but sstabledump gives no result at all. (on all nodes).

It is ok. You can give nodetool getendpoints any key (whether it's present in data or not) and it will tell you where it belongs to.

**Comment by Termite** [comment]
@haaak: 
There was not the slightest change to the cluster

Ok, I see, the destination nodes are just calculated from the key, or the murmured key

**Comment by haaawk** [comment]
@Termite Could you please give an example of a key that has a mismatch between nodetool getendpoints and actual nodes with sstables that contain them?

It would be great if you could describe your cluster and give an example of a key with a mismatch describing which nodes are returned from nodetool getendpoint for that key and which nodes actually contain sstables with that key.

**Comment by Termite** [comment]
Yes of course, I can give you some keys which have a mismatch.

The cluster is the same as described at the beginning of this thread

**Comment by eliransin** [comment]
I am still bisecting (it will take a half a day more) but after the last few rounds the suspect list is ~50 commits with a one very promising candidate and a few others that are suspects just because they 
have sstable in their name:
Main suspect: 6c5f8e0ed sstables: index_reader: Simplify offset calculations

Full suspect list:
0eb703dc8 all: rename view_update_from_staging_generator
a5d24e40e distributed_loader: fix indentation
13c8c8404 service: add generating view updates from uploaded sstables
46305861c init: pass view update generator to storage service
13f645335 sstables: treat sstables in upload dir as needing view build
09401e0e7 sstables,table: rename is_staging to requires_view_building
76616f680 distributed_loader: use proper directory for opening SSTable
beb483672 db,view: make throttling optional for view_update_generator
906275008 scylla_util.py: make view_hints_directory setting optional
238866228 memtable: rename get_stats to get_encoding_stats
df090a15f Merge "Add counters for inactive reads" from Botond
acd6999ba Don't use SEASTAR_HAVE_LZ4_COMPRESS_DEFAULT in scylla
684fb607c sstable: handle missing index entry
f5ee466a1 Merge "Cleanup UDT and tuple names creation" from Piotr
f12a3e206 sstables: index_reader: Rename _promoted_index_size
6c5f8e0ed sstables: index_reader: Simplify offset calculations
0d52bdcba install-dependencies.sh: unwrap long lines
391d1e0fe table: const correctness for table::get_sstables() and related
cd9ce1887 sstable: rename the is_boundary predicate
96b880f81 Add comment explaining tuple type name creation
57e655d71 Add "FrozenType(...)" to UDT name only when it's frozen
fc17bd376 Move "FrozenType(...)" addition to UDT name to user_type_impl
1fdfc461b Add "frozen<...>" to tuple CQL name only when it's frozen
749eee271 Move "frozen<...>" addition to tuple CQL name to tuple_type_impl
7aba17de2 Merge make_cql3_tuple_type into tuple_type_impl::as_cql3_type
56060573b Add "frozen<...>" to UDT CQL name only when it's frozen
a928c103c Move "frozen<...>" addition to UDT CQL name to user_type_impl
1b7cad353 database: Fix race condition in sstable snapshot
2dc377640 sstables: mc: sign-extend serialization_header min_local_deletion_time_base and min_ttl_base
a29182b44 sstable: fix use after free while applying extensions in sstable::open_file
b247ce01c table: restore indentation after changes to table::make_sstable_reader
3d6be2f82 table: reduce duplication in table::make_sstable_reader
16dda033a sstables: row_marker: initialize _expiry
4a6aeced5 Merge "Fix UDTs representation in serialization header" from Piotr
d4bc5b64c Add test for serialization header with UDT
3de85aebc Fix UDT names in serialization header
60323b79d sstables: mc: sign-extend delta local_deletion_time and delta ttl
26ac2c23e Change *_row_* names that refer to partitions
f00e9051e reloc: show error message when relocatable package doesn't exist
f5301990f compaction: release reference of cleaned sstable in compaction manager
fa2b0384d Replace std::experimental types with C++17 std version.
51a08c324 sstable: remove constexpr from run time predicates
c5346cdf9 database, table: split table-related code to table.cc
8ecb528d5 Update seastar submodule
d8adbeda1 tests: mutation_source_test: generate valid utf-8 data
1de24c849 repair: Use mf.visit() in fragment_hasher
67039e942 Remove the only use of with_alignment from scylla
0d4529a5f Change timeout to fix tests in a debug build
34251f5ea tools: toolchain: update image for all-user sudo
80affca5f tests/querier_cache: use stats to get the no. of inactive reads
e56c26205 reader_concurrency_semaphore: add counters for inactive reads

Will update as the list gets shorter :slightly_smiling_face: 
/cc @bhalevy @slivne

**Comment by eliransin** [comment]
BTW - maybe it only happens when the index size is too big and thus the math in the above commit gets wrong. A link to an example index file (with the infamous 77 key can be found) [here](https://github.com/scylladb/scylla/issues/6040#issuecomment-630120303)

**Comment by slivne** [comment]
lets split the issues out - nodetool getendpoints seems to have a bug -
take this out of this issue.

please compare nodetool getendpoints for an incorrect key and what replicas
are shown when you do a
cqlsh>
tracing on;
select * from ....

we will try to reproduce locally - yet nodetool getendpoints - should be
split to a new issue.

**Comment by eliransin** [comment]
> please compare nodetool getendpoints for an incorrect key and what replicas are shown when you do a cqlsh> tracing on; select * from ....

Already did, the coordinator queries the right endpoints which are different than the ones we get through nodetool/rest api.

**Comment by slivne** [comment]
@Eliran Sinvani <eliransin@scylladb.com> thanks for the confirmation - so
lets split the nodetool getendpoints to a different issue - and have this
one for the missing keys.

**Comment by eliransin** [comment]
OK, this hunch made things faster, found it :slightly_smiling_face: 
I don't know enough about this area of the code to just see if the problem is in position() function 
that was introduced in another commit or in the new position calculations introduced in this one.

```
[eliransin@localhost scylla]$ git bisect bad
6c5f8e0eda61c8010675b46a9a3beb51b563a29a is the first bad commit
commit 6c5f8e0eda61c8010675b46a9a3beb51b563a29a
Author: Tomasz Grabiec <tgrabiec@scylladb.com>
Date:   Fri Jan 11 16:07:13 2019 +0100

    sstables: index_reader: Simplify offset calculations
    
    Now that continuous_data_consumer::position() is meaningful (since
    36dd660), we can use our position in the stream to calculate offsets
    instead of duplicating state machine in offset calculations.
    
    The value of position() - data.size() always holds the current offset
    in the stream.
    Message-Id: <1547219234-21182-1-git-send-email-tgrabiec@scylladb.com>

 sstables/index_reader.hh | 28 ++++++++++------------------
 1 file changed, 10 insertions(+), 18 deletions(-)
```
/cc @tgrabiec @bhalevy 

a very simple reproducer:

1.  Download the sampled example ( `gsutil cp gs://upload.scylladb.com/7cf8c51a-8ab4-4798-9248-71dcb754bb66/* .`) into some directory **dir**
2. buid your version with dbuild
3. from the scylla source dir run: `sudo tools/toolchain/dbuild -ti --network bridge -v <dir>:<dir> -- /bin/bash`  - things to notice, the use of **sudo**, the use of **network** and the use of **-v** to make the sample sstable available and that it is not detached.
4. in this docker start scylla by running : `build/release/scylla  --developer-mode=1  --smp 1 --overprovisioned --listen-address 127.0.0.1 --rpc-address 127.0.0.1 --seed-provider-parameters seeds=127.0.0.1 --blocked-reactor-notify-ms 999999999`
5. in another shell (or shell tab) , and assuming the container from the previous step is called **docker1** (which is much more original from docker's generated names :slightly_smiling_face: ) run the following: `docker exec -ti docker1 /bin/bash` (lets assume this is shell2)
6. inside shell 2 (which is connected to docker) run: `pip install cqlsh`
7. `cqlsh`
8. `CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true ;`
9. `CREATE TABLE test.pk_only4 (pk bigint PRIMARY KEY) WITH compaction = {'class': 'NullCompactionStrategy'}; `
10. `quit` to close cqlsh (but not the docker)
11. go back to docker1 (in shell1 where scylla is running) and press `Ctrl+C` to shudtdown scylla
12. in either shells (inside docker) cd into the tables dir : `/var/lib/scylla/data/test/pk_only4???`
13. ` cp <dir>/mc* .` 
14. from the first shell run the command from step 4 again in order to start scylla
15. in the second shell : `cqlsh`
16. `SELECT * FROM test.pk_only4 LIMIT 10;` - just to see that the table is actually loaded.
17. Reproducer: `SELECT * FROM test.pk_only4 WHERE pk = 77;` , this will return no rows for
77 which is in this sstable.

Let  me know if you nee anything else.

**Comment by avikivity** [comment]
Backported to 3.3, 4.0, 4.1.
